Make kube-proxy iptables SNAT masquerade mark configurable

matthewdupre opened the pull request:
Fixes issue #19983.  We'd love to get this into 1.2.

The kube-proxy currently uses a fixed iptables mark, which means no other application can use the mark (which has built in support for masking and only using part of the space).  This fix adds the ability to configure the mark into kube-proxy, as well as changing the default to only use part of the space.

matthewdupre commented:
OK, I've been added to the corporate CLA now (company: Metaswitch Networks, google account matthew.dupre@metaswitch.com).

matthewdupre commented:
I'm working on the rebase.  It looks like #18600 has changed the configuration mechanism: proxy-mode now defaults to userspace not iptables, and is no longer specifiable on the command line invocation of kube-proxy (or at least, it's ignored).  As a result I'm still working on doing a test of the rebased code.

thockin commented:
The change in the default is a bug and the fix is in flight.

On Tue, Feb 2, 2016 at 12:33 PM, Matt Dupre <notifications@github.com>
wrote:

> I'm working on the rebase. It looks like #18600
> <https://github.com/kubernetes/kubernetes/pull/18600> has changed the
> configuration mechanism: proxy-mode now defaults to userspace not iptables,
> and is no longer specifiable on the command line invocation of kube-proxy
> (or at least, it's ignored). As a result I'm still working on doing a test
> of the rebased code.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20496#issuecomment-178807395>
> .
>

thockin commented on cmd/kube-proxy/app/options/options.go:
```diff
@@ -64,6 +65,7 @@ func NewProxyConfig() *ProxyServerConfig {
 		HealthzBindAddress:             net.ParseIP("127.0.0.1"),
 		OOMScoreAdj:                    qos.KubeProxyOOMScoreAdj,
 		ResourceContainer:              "/kube-proxy",
+		IptablesMasqueradeMark:         "0x4d410000/0xffff0000",
```
why not just make it a single bit and just make the option be the bit number [0-31] ?

matthewdupre replied:
Sure.  Call it iptablesMasqueradeBit?

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -181,12 +180,13 @@ func NewProxier(ipt utiliptables.Interface, exec utilexec.Interface, syncPeriod
+		syncPeriod:             syncPeriod,
+		iptables:               ipt,
+		masqueradeAll:          masqueradeAll,
+		iptablesMasqueradeMark: iptablesMasqueradeMark,
```
the 'iptables' prefix here is unnecessary

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -204,11 +204,8 @@ func CleanupLeftovers(ipt utiliptables.Interface) (encounteredError bool) {
-		encounteredError = true
-	}
+	// remove all masquerade rules from the POSTROUTING chain.
+	encounteredError = ensureMasqueradeRuleOnly(ipt, "") || encounteredError
```
Ohh, good catch.  I think this should be rejiggered to have a single jump rule in POSTROUTING to a KUBE-MASQ chain.  Then cleanup is a single unqualified delete from POSTROUTING and a flush and remove of KUBE-MASQ.

If I am asking too much, say so, and we can do that as a followup

thockin replied:
I misinterpreted what this was doing.  I'll have to process this more deeply

matthewdupre replied:
Most of this effort is dealing with old rules floating around for outdated marks.  A single chain jump is a good option: I haven't done it (yet?) because I wasn't sure I could get it done in time, and even with the chain the code to delete the original mark rule still has to stay around to allow people to upgrade without flushing iptables.

Let me know what you think.

thockin replied:
I think we can still do it this way, but the transition is annoying.  Making the change to a new-chain means that we can eventually delete the transition, though.

How about this.  Unilaterally delete the old-mark rule, suppress errors, add a TODO to remove that block after 1.2 is cut.  Add a KUBE-APPLY-MASQ chain, and a unilateral jump from POSTROUTING to KUBE-APPLY-MASQ.  And appropriate cleanup.  Then on startup we can simply save and restore that chain with the new value.

Oh puke, this means all the rules that set the mark ALSO need to change on every kube-proxy restart.  Blech.

We could do a similar thing - have a KUBE-SET-MASQ chain that sets the marks, so we can save/restore and make all the KUBE-SEP chains jump to that rather than setting the mark directly.

Again, the transition is annoying, but we get to delete it.

Am I asking too much?  This is harder than I expected (I'm going to have that written on my gravestone, I think)

thockin commented:
I just stared at the syncProxyRules() logic and, while it is not a pretty change, it does not seem TOO bad.  We already rebuild all of the per-endpoint and per-service rules anyway. If you need help, I can pre-flight the basic moving of things for you.  I try to not abuse contributors, if I can avoid it, but this got a little bigger than expected.

Let me know ASAP if you want me to help - it's going to be a busy week and there are at least 2 other PRs this might conflict with :)

thockin commented:
#19576 is the one that will cause rebase pain for this PR - might want to rebase onto it to start.

matthewdupre commented:
Thanks for the prompt and useful feedback @thockin.

So you're suggesting I base on top of the #19576 change?  I don't want to jeopardize hitting 1.2 if possible, but if that PR will make it in then no problem.

I'll take a look at the actual logic tomorrow morning when I'm fresher.

thockin commented:
#19576 will make it for 1.2 and you'll have to rebase onto it sooner or later. Sooner is easier.

I might take a look at pre-flighting some of the changes - I don't want to send you on a snipe hunt.  If I do, I'll send you a PR and you can either rebase onto THAT or else rip off my changes directly.

thockin committed:
Dont log errors on proxy leftover cleanup

thockin committed:
Put all masquerade mark logic into new chains

This allows us to use the MARK-MASQ chain as a subroutine, rather than encoding
the mark in many places.  Having a KUBE-POSTROUTING chain means we can flush
and rebuild it atomically.  This makes followon work to change the mark
significantly easier.

thockin committed:
Implement proper cleanup in iptables proxy

thockin commented:
I cleared the road for you in #20541

matthewdupre commented:
@thockin: thanks for your help with this.

I've took a suitable starting point (a recent master with #20541, (therefore #19576) and #20515), and rebased this change on top of it.  I've tried to make sure your feedback was incorporated, although it does now look quite different (mostly I've deleted a bunch of code thanks to your KUBE-POSTROUTING work).

matthewdupre commented:
OK, I've rebased onto your rebased PR.  Still pulling in Mike's #20515.

matthewdupre commented:
@thockin: I've rebased onto a recent master (had conflicts in some generated files), been through all the autogenerators (had some trouble with update-codecgen.sh, but worked around it) and ran everything through verify-all.sh.

Hopefully the PR is now in a good state for review / merge, so over to you for that.  Thanks, Matt

matthewdupre commented:
@thockin: can I get this ran through the tests please (I think they need to be triggered?).

thockin commented:
ok to test

thockin commented:
Needs a rebase and I had a nit or two, otherwise LGTM.

Please squash your commits together and we can get this and the one it deps on into the queue.


Thanks!

thockin commented:
ping me here when you;re done

caseydavenport commented:
@thockin - It's pretty late in the UK right now so I've tried to take over where @matthewdupre left off.  I've opened a new squashed PR with your comments addressed.

https://github.com/kubernetes/kubernetes/pull/20768

thockin commented:
If you're in a panic because of code freeze, don't sweat that - this is "substantially done".  It's not QUITE like the airport, where the door closes and that is that.  The goal is to stop new features while we drain the queue of things that are really pretty much done.

Do you want me to stay on here or review the new branch?

matthewdupre commented:
Thanks @thockin: I've picked this back up, and sorted out the issues with #20768.  This branch now contains the latest work on this.

thockin commented on cmd/kube-proxy/app/server.go:
```diff
@@ -190,18 +190,23 @@ func NewProxyServerDefault(config *options.ProxyServerConfig) (*ProxyServer, err
 		endpointsHandler = proxierIptables
 		// No turning back. Remove artifacts that might still exist from the userspace Proxier.
-		glog.V(2).Info("Tearing down userspace rules. Errors here are acceptable.")
+		glog.V(0).Info("Tearing down userspace rules.")
```
This should stay V(2)

thockin commented on cmd/kube-proxy/app/server.go:
```diff
@@ -221,7 +226,7 @@ func NewProxyServerDefault(config *options.ProxyServerConfig) (*ProxyServer, err
 		proxier = proxierUserspace
 		// Remove artifacts from the pure-iptables Proxier.
-		glog.V(2).Info("Tearing down pure-iptables proxy rules. Errors here are acceptable.")
+		glog.V(0).Info("Tearing down pure-iptables proxy rules.")
```
should stay V(2) I think

thockin commented on pkg/apis/componentconfig/types.go:
```diff
@@ -31,6 +31,9 @@ type KubeProxyConfiguration struct {
 	HostnameOverride string `json:"hostnameOverride"`
+	// iptablesMasqueradeBit is the bit of the iptables fwmark space to use for SNAT if using
+	// the pure iptables proxy mode. Values must be within the range [0, 31].
+	IPTablesMasqueradeBit *int `json:"iptablesMasqueradeBit"`
```
I think you can make the internal (non-versioned) type use an int, and the versioned type use a pointer, and that way the code that actually uses the internal type doesn't have to much with dereferencing and defaulting.  The smell here is that the value `14` turns up outside of the defaults.go file at all.

thockin commented:
small nits and squash please.

thockin commented:
add to whitelist

matthewdupre commented:
How far down do you want it squashing?  Your 3 commits + change + autogen (so 5 total)?

thockin commented:
squash just your two, since they are intrinsically one piece.

On Sun, Feb 7, 2016 at 1:47 PM, Matt Dupre <notifications@github.com> wrote:

> How far down do you want it squashing? Your 3 commits + change + autogen
> (so 5 total)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20496#issuecomment-181125920>
> .
>

matthewdupre commented:
Are you sure about the v(2)/v(0) ones (given it was your change to begin with!)?

thockin commented:
LOL, sorry, I started reviewing the full diff and then switched to just
your commits.  Yeah, don't "fix" that, I guess.

On Sun, Feb 7, 2016 at 1:50 PM, Matt Dupre <notifications@github.com> wrote:

> Are you sure about the v(2)/v(0) ones (given it was your change to begin
> with!)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20496#issuecomment-181128221>
> .
>

matthewdupre commented:
I had a look at the int * / too many 14s comment: it looks like the defaulting in server.go is actually superfluous (I had a bit of trouble getting the pointer version to work until I released it actually depends on the autogenerated stuff, so added that check in to try and fix it).

That'll bring it down to just in options.go and defaults.go.  To be honest I'm not clear on where the APIs  actually fall: I'm not really familiar with the Kubernetes code, so I've mostly just mirrored the equivalent code for OOMScoreAdj.  If you can suggest a good home for a constant I can definitely add that: otherwise I'd probably suggest keeping it the same as OOMScoreAdj.  Does that sound OK?

thockin commented:
The config is stored as versioned (which you can see in the path to the types.go) but converted to the internal version (the types.go with no version).  In conversion it applies default values, wherein you correctly default to 14.  The problem comes in the consumer.  By leaving it as a pointer. you REALLY need to be checking for nil, and in this case you have to do something sane if it is nil.  You could return an error, I guess, but I think it's cleaner all around to have the non-versioned type use a non-pointer.  Then you don't have to check for nil, and tests which end up passing a config in, but don't care about the mas

thockin commented:
The versioned API (as denoted by the v1alpha1 in the path tp types.go) is what
is used to store the config, but th einternal (no version) struct is what is
used by the app internally.

You correctly default nil to 14 in defaults.go, so the internal code shouldn't
have to worry about nil.  But it turns out that tests and other things end up
passing half-baked Configs in (why set the masq bit when the test doesn't test
that).  This means you really ought to test for nil.

You could return an error (which I guess is OK) or you could convert to a
non-pointer for the internal API.  I think th eauto-generators will not make
this pleasant, which will mean you have to hand0write the conversion, which is
unpleasant and not something we want more of.

I think the trend in the medium term is to move away from unversioned types (@lavalamp, @mikedanese),
which leaves us with the nil problem.  I guess you can just check for it and
return an error.

matthewdupre commented:
OK; how's this then: I'll change the test in server.go (the only time I actually pull the option out of the ProxyServerConfig) that currently sets the default if it's nil, to actually just error out?

I think that plus squashing the autogenerator commit in covers everything?

thockin commented:
I think so.

On Sun, Feb 7, 2016 at 2:50 PM, Matt Dupre <notifications@github.com> wrote:

> OK; how's this then: I'll change the test in server.go (the only time I
> actually pull the option out of the ProxyServerConfig) that currently sets
> the default if it's nil, to actually just error out?
>
> I think that plus squashing the autogenerator commit in covers everything?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20496#issuecomment-181137116>
> .
>

matthewdupre committed:
Change iptables fwmark to use single configurable bit instead of whole mark space

matthewdupre commented:
OK, markups made @thockin.

thockin commented:
LGTM!
